packages/flutter/lib/src/material/app_bar.dart

```diff
@@ -1254,14 +1254,20 @@ class _SliverAppBarDelegate extends SliverPersistentHeaderDelegate {
         child: title,
       ),
     };
+    final ScaffoldState? scaffold = Scaffold.maybeOf(context);
+    final ModalRoute<dynamic>? parentRoute = ModalRoute.of(context);
+    bool hasLeading = leading != null;
+    if (leading == null && automaticallyImplyLeading) {
+      hasLeading = (scaffold?.hasDrawer ?? false) || (parentRoute?.impliesAppBarDismissal ?? false);
+    }
```

Comment:
I've probably spent too much time seeing how this could be reworked. I've tripped over it though every time I've read it. 😆 

```suggestion
    bool hasLeading = leading != null || automaticallyImplyLeading;
    if (!hasLeading) {
      hasLeading = (scaffold?.hasDrawer ?? false) || (parentRoute?.impliesAppBarDismissal ?? false);
    }
```

Reply:
Nice catch!